src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -0,0 +1,563 @@
+        private StartState _started;
+
+        // Used by the class to indicate that the stream is m_Readable.
+        private bool _canRead;
+
+        // Used by the class to indicate that the stream is writable.
+        private bool _canWrite;
+
+        // Pipe for reading.
+        private Pipe _requestPipe;
+
+        private volatile bool _disposed = false;
+
+        // Creates a new MsQuicStream
+        internal MsQuicStream(MsQuicApi api, MsQuicConnection connection, QUIC_STREAM_OPEN_FLAG flags, IntPtr nativeObjPtr, bool inbound)
+        {
+            Debug.Assert(connection != null);
+
+            _api = api;
+            _ptr = nativeObjPtr;
+
+            if (inbound)
+            {
+                _started = StartState.Finished;
+
+                _canWrite = !flags.HasFlag(QUIC_STREAM_OPEN_FLAG.UNIDIRECTIONAL);
+                _canRead = true;
+            }
+            else
+            {
+                _started = StartState.None;
+
+                _canWrite = true;
+                _canRead = !flags.HasFlag(QUIC_STREAM_OPEN_FLAG.UNIDIRECTIONAL);
+            }
+
+            // Options to effectively have a single buffer pipe
+            // Effectively this is a synchronization mechanism between calls to ReadAsync and Recv callbacks
+            // However, having a pipe here is nice for calling Complete,
+            // handing buffering, and handling synchronization.
+            var inputOptions = new PipeOptions(pool: null, PipeScheduler.ThreadPool, PipeScheduler.Inline, pauseWriterThreshold: 1, resumeWriterThreshold: 1, useSynchronizationContext: false);
+
+            _requestPipe = new Pipe(inputOptions);
+            _sendResettableCompletionSource = new UIntResettableCompletionSource(this);
+            SetCallbackHandler();
+        }
+
+        internal override bool CanRead => _canRead;
+
+        internal override bool CanWrite => _canWrite;
+
+        internal override long StreamId
+        {
+            get
+            {
+                if (_streamId == -1)
+                {
+                    _streamId = GetStreamId();
+                }
+
+                return _streamId;
+            }
+        }
+
+        internal override async ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            if (_started == StartState.None)
+            {
+                _started = StartState.Started;
+                await StartAsync();
+            }
+
+            await SendAsync(new ReadOnlySequence<byte>(buffer), QUIC_SEND_FLAG.NONE);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+        }
+
+        internal override async ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            while (true)
```

Comment:
Does this need a while true?

Reply:
depends on if I call CancelPendingRead.

Reply:
I don't think I need to call CancelPendingRead for now, so removing.